### ISSUE TITLE
LLaMA Factory 사용 방법 숙지 및 학습 시작

### ISSUE THREAD

**Issue by ryangkyung-thaki** [ThakiCloud/LLaMA-Factory#1]
## 📚 목표

LLaMA Factory를 활용한 대규모 언어 모델(LLM) fine-tuning 방법을 학습하고 실습합니다.

### 1. 기본 개념 이해
- [x] LLaMA Factory 프레임워크 개요 파악
- [x] 지원되는 모델 및 training 방법 확인 (LoRA, QLoRA, Full-tuning 등)
- [ ] 하드웨어 요구사항 검토

### 2. 환경 설정
- [x] 의존성 설치 및 환경 구성
- [x] GPU/NPU 환경 확인
- [x] 필요한 라이브러리 설치 (torch, transformers, peft, trl 등)

### 3. 데이터 준비
- [x] 데이터셋 포맷 이해 (`data/README.md` 참고)
- [x] `data/dataset_info.json` 설정 방법 학습
- [x] 제공되는 데이터셋 확인 및 커스텀 데이터셋 작성 방법 학습

### 4. 모델 학습 실습
- [x] CLI를 통한 기본 LoRA fine-tuning 실행
  - 예시: `llamafactory-cli train examples/train_lora/llama3_lora_sft.yaml`
- [x] YAML 설정 파일 구조 이해
- [ ] 학습 파라미터 조정 (learning_rate, batch_size, epochs 등)
- [ ] Web UI (LLaMA Board) 사용해보기

### 5. 모델 추론 및 배포
- [ ] 학습된 모델로 추론하기
  - 예시: `llamafactory-cli chat examples/inference/llama3_lora_sft.yaml`
- [ ] LoRA 가중치 병합하기
  - 예시: `llamafactory-cli export examples/merge_lora/llama3_lora_sft.yaml`
- [ ] vLLM을 활용한 OpenAI 스타일 API 배포

## 📖 참고 자료

- **공식 문서**: https://llamafactory.readthedocs.io/en/latest/
- **GitHub 저장소**: https://github.com/hiyouga/LLaMA-Factory
- **예제 파일**: `examples/` 디렉토리
- **Colab 노트북**: https://colab.research.google.com/drive/1eRTPn37ltBbYsISy9Aw2NuI2Aq5CQrD9?usp=sharing
- **FAQ**: https://github.com/hiyouga/LLaMA-Factory/issues/4614

**Comment by ryangkyung-thaki** [comment]
CUDA Out Of Memory

모델 로딩 단계에서 실패했습니다. 즉:
❌ LoRA adapter 생성 전
❌ Gradient checkpointing 적용 전
❌ 학습 시작 전

```
Loading checkpoint shards: 45% | 21/47
                          ↑
                 모델 로딩 중에 OOM!
```

🔍 이론 vs 실제
- 이론적 계산
   - 모델 크기: 212GB (BF16-default 모델 로드 설정)
   - GPU 4개로 분산: 212GB / 4 = 53GB per GPU
   - 로딩만 할 때: 53GB per GPU만 필요
   - 사용 가능: 80GB per GPU
   - → 충분해야 함! ✅

🤔 왜 이런 일이 발생하나?
1. Transformers의 비효율적인 로딩 방식
- from_pretrained()는 모델을 로드할 때: 
   - 이론적으로 기대하는 것: 각 GPU에 자기 몫(53GB)만 로드
   - 실제로 일어나는 것:
      1. 전체 모델을 메모리에 로드 (일시적)
      2. 그 다음 각 GPU로 분산
      3. 중간 과정에서 중복 메모리 사용

2. Sharded Checkpoint의 문제
- 47개의 파일을 순차적으로 로드하면서:
   - 일부는 이미 GPU에 올라가 있고
   - 새로운 shard를 로드하려고 하면
   - 피크 메모리 사용량이 이론값보다 훨씬 높아짐

FP8 모델 사용??
https://huggingface.co/zai-org/GLM-4.5-Air-FP8

**Comment by ryangkyung-thaki** [comment]
## 문제 원인
- 현재 DDP (DistributedDataParallel) 방식으로 학습하고 있는데, DDP는 각 GPU에 모델 전체를 복사합니다. 
- GLM-4.5-Air는 MoE(Mixture of Experts) 모델로 매우 크기 때문에(FP8 양자화 후에도), 각 GPU에 전체 모델을 로드하면 메모리가 부족합니다.
- 해결 방법: DeepSpeed ZeRO Stage 3 사용
- 모델 파라미터를 여러 GPU에 샤딩하여 메모리를 분산시켜야 합니다.
